VPD parser crashes in some cases

# Required information

 * Distribution: Ubuntu
 * Distribution version: 22.04 (not specific to distro)
   * Kernel version: 5.15.0-40-generic (not specific to this kernel)
   * LXC version: 5.0.0
   * LXD version: 5.3

# Issue description

MAAS is relying on LXD codebase for discovery of machine resources and I found an issue after upgrading MAAS to a newer version (3.2) which utilizes the machine resources binaries which have the VPD parser included: https://github.com/lxc/lxd/commit/6ccaead4469927db496ec6ca808b06c1f78fc718

The crash comes as a result of a panic due to going outside of a slice range. The issue is reproducible when the tool is run in isolation as well.

```
        provisioningserver.utils.shell.ExternalProcessError: Command `sudo /usr/share/maas/machine-resources/amd64` returned non-zero exit status 2:                                                                                                                                    
        panic: runtime error: slice bounds out of range [65535:32611]                                                                                                                                                                                                                   
        goroutine 1 [running]:                                                                                                                      github.com/lxc/lxd/lxd/resources.parsePCIVPD({0xc000d06000, 0x8000, 0x8001})                                                        
                /build/maas-QVndNG/maas-3.2.0-11989-g.84a255c14/src/host-info/vendor/github.com/lxc/lxd/lxd/resources/pci_vpd.go:116 +0x312 
        github.com/lxc/lxd/lxd/resources.GetPCI()
                /build/maas-QVndNG/maas-3.2.0-11989-g.84a255c14/src/host-info/vendor/github.com/lxc/lxd/lxd/resources/pci.go:146 +0x9e5
        github.com/lxc/lxd/lxd/resources.GetResources()
                /build/maas-QVndNG/maas-3.2.0-11989-g.84a255c14/src/host-info/vendor/github.com/lxc/lxd/lxd/resources/resources.go:48 +0x267
        host-info/pkg/info.GetResources(...)
                /build/maas-QVndNG/maas-3.2.0-11989-g.84a255c14/src/host-info/pkg/info/info.go:157
        host-info/pkg/info.GetInfo()
                /build/maas-QVndNG/maas-3.2.0-11989-g.84a255c14/src/host-info/pkg/info/info.go:182 +0x34
        main.main()
                /build/maas-QVndNG/maas-3.2.0-11989-g.84a255c14/src/host-info/cmd/machine-resources/main.go:23 +0x1d
```

Stracing have shown that it is specific to some VPD entries only and which `/sys/bus/pci/devices/dddd:bb:dd.f/vpd` paths to look at.

I bind mounted an empty file on top of each entry and managed to avoid the crash in the end:

```
touch fakevpd
sudo mount --bind fakevpd /sys/bus/pci/devices/0000:01:00.0/vpd
```

The VPD blob is exposed by the kernel since v2.6.26 https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=94e6108803469a37ee1e3c92dafdd1d59298602f and has a format defined in specs dated back to PCI 2.2 and has not seen change since then (but got included verbatim into PCIe 4.0).

I attached a couple of examples of VPDs that make the parser crash in an archive here (blobs as exposed by the kernel in the PCI VPD spec format).
[bad-vpds.zip](https://github.com/lxc/lxd/files/9028036/bad-vpds.zip)

`lspci -vvv` is able to handle those correctly and print something like this:

```
        Capabilities: [e0] Vital Product Data
                Product Name: X520 10GbE Controller
                Read-only fields:
                        [PN] Part number: G61346
                        [MN] Manufacture ID: 1028
                        [V0] Vendor specific: FFV17.5.10
                        [V1] Vendor specific: DSV1028VPDR.VER1.0
                        [V3] Vendor specific: DTINIC
                        [V4] Vendor specific: DCM10010081D521010081D5
                        [V5] Vendor specific: NPY2
                        [V6] Vendor specific: PMT12345678
                        [V7] Vendor specific: NMVIntel Corp
                        [RV] Reserved: checksum good, 0 byte(s) reserved
                End
```

Vendors do not always fully conform to the spec so there are some invalid cases that should be handled gracefully and not make the parser crash at least. In the Libvirt code base the approach was also taken to handle some cases gracefully where most of the VPD is valid but some fields use non-standard approaches.

https://gitlab.com/libvirt/libvirt/-/blob/v8.5.0/src/util/virpcivpd.c#L654 (parser)
https://gitlab.com/libvirt/libvirt/-/blob/v8.5.0/tests/virpcivpdtest.c#L863-983 (tests for various invalid and valid cases).

I'm also seeing this - 
* Distribution: Ubuntu
* Distribution version: 22.04 (not specific to distro)
* Kernel version: 5.15.0-40-generic (not specific to this kernel)
    * LXC version 4.0.9
    * LXD version  4.0.9

```
panic: runtime error: slice bounds out of range [1:0]

goroutine 1 [running]:
github.com/lxc/lxd/lxd/resources.vpdReadInt(...)
	/build/maas/parts/host-info/build/vendor/github.com/lxc/lxd/lxd/resources/pci_vpd.go:42
github.com/lxc/lxd/lxd/resources.vpdReadEntries({0xc0001bd546, 0xfba, 0x14ba}, 0xaa)
	/build/maas/parts/host-info/build/vendor/github.com/lxc/lxd/lxd/resources/pci_vpd.go:66 +0x337
github.com/lxc/lxd/lxd/resources.parsePCIVPD({0xc0001bd500, 0x1000, 0x1500})
	/build/maas/parts/host-info/build/vendor/github.com/lxc/lxd/lxd/resources/pci_vpd.go:108 +0x127
github.com/lxc/lxd/lxd/resources.GetPCI()
	/build/maas/parts/host-info/build/vendor/github.com/lxc/lxd/lxd/resources/pci.go:146 +0x9e5
github.com/lxc/lxd/lxd/resources.GetResources()
	/build/maas/parts/host-info/build/vendor/github.com/lxc/lxd/lxd/resources/resources.go:48 +0x267
host-info/pkg/info.GetResources(...)
	/build/maas/parts/host-info/build/pkg/info/info.go:157
host-info/pkg/info.GetInfo()
	/build/maas/parts/host-info/build/pkg/info/info.go:182 +0x34
main.main()
	/build/maas/parts/host-info/build/cmd/machine-resources/main.go:23 +0x1d
```

I'll add my examples once I have located the failing VPD data.

--- a/lxd/resources/pci_vpd.go
+++ b/lxd/resources/pci_vpd.go
@@ -115,6 +115,11 @@ func parsePCIVPD(buf []byte) api.ResourcesPCIVPD {
 			}
 
 		default:
+			// Check that we aren't past the buffer (invalid length).
+			if len(buf) < length {
+				break
+			}
+
 			// For other tags, just skip the value.
 			buf = buf[length:]
 		}

